src/analysis_and_optimization/Mem_pattern.ml

```diff
@@ -0,0 +1,602 @@
+  | MultiIndex exprs -> op exprs
+
+(**
+ * Search through an expression for `Var name` where `name` is in
+ *  the `aos_exits`
+ * and the type of the expression contains an eigen type.
+ * @param aos_exits A Set of strings with the name of the obj
+ *  we are searching for.
+ * @param pattern An expression pattern we recursivly search through 
+ *  for the name.
+ **)
+let rec query_expr_eigen_names (aos_exits : string Set.Poly.t)
+    Expr.Fixed.({pattern; meta= Expr.Typed.Meta.({type_; _})}) =
+  let query_name = query_expr_eigen_names aos_exits in
+  match pattern with
+  | FunApp ((_ : Fun_kind.t), (exprs : Expr.Typed.Meta.t Expr.Fixed.t list)) ->
+      List.exists ~f:query_name exprs
+  | TernaryIf (predicate, texpr, fexpr) ->
+      query_name predicate || query_name texpr || query_name fexpr
+  | Indexed (expr, indexed) ->
+      query_name expr
+      || List.exists ~f:(name_exists_in_index query_name) indexed
+  | Var (name : string) when UnsizedType.contains_eigen_type type_ ->
+      Set.Poly.mem aos_exits name
+  | Var (_ : string) -> false
+  | Lit ((_ : Expr.Fixed.Pattern.litType), (_ : string)) -> false
+  | EAnd (lhs, rhs) | EOr (lhs, rhs) -> query_name lhs || query_name rhs
+
+(**
+ * Search through an expression for `Var name` where `name` is in
+ *  the `aos_exits` and return `true` if it exists and false otherwise.
+ * and the type of the expression contains an eigen type.
+ * @param aos_exits A Set of strings with the name of the obj
+ *  we are searching for.
+ * @param pattern An expression pattern we recursivly search through 
+ *  for the name.
+ **)
+let rec query_all_expr_eigen_names
+    Expr.Fixed.({pattern; meta= Expr.Typed.Meta.({type_; _})}) =
+  let find_name = query_all_expr_eigen_names in
+  match pattern with
+  | FunApp ((_ : Fun_kind.t), (exprs : Expr.Typed.Meta.t Expr.Fixed.t list)) ->
+      Set.Poly.union_list (List.map ~f:find_name exprs)
+  | TernaryIf (predicate, texpr, fexpr) ->
+      Set.Poly.union_list
+        [find_name predicate; find_name texpr; find_name fexpr]
+  | Indexed (expr, indexed) ->
+      let find_in_index = map_index find_name in
+      Set.Poly.union (find_name expr)
+        (Set.Poly.union_list (List.map ~f:find_in_index indexed))
+  | Var (name : string) when UnsizedType.contains_eigen_type type_ ->
+      Set.Poly.singleton name
+  | Var (_ : string) -> Set.Poly.empty
+  | Lit ((_ : Expr.Fixed.Pattern.litType), (_ : string)) -> Set.Poly.empty
+  | EAnd (lhs, rhs) | EOr (lhs, rhs) ->
+      Set.Poly.union (find_name lhs) (find_name rhs)
+
+(** 
+ * Modify functions expressions from SoA to AoS
+ * The only real path in the below is on the functions.
+ * If all of the eigen types in the function's expression list
+ * We make the function AoS, else it stays as SoA.
+ *
+ * @param aos_exits The name of the variable whose
+ *  associated expressions we want to modify.
+ * @param expr The expression to modify
+ *)
+let rec demote_exprs (aos_exits : string Set.Poly.t)
```

Comment:
I'd suggest implementing this with Mir_utils.map_rec_expr. You just pass it a pattern -> pattern function and it changes all the subexpressions, so you can avoid rewriting the recursion logic

Reply:
Done!